# Homework Assignment #4 (Individual)
## Regression models

### <p style="text-align: right;"> &#9989; Rajmeet Singh Chandok</p>
### <p style="text-align: right;"> &#9989; rajmeet</p>

### Goal for this homework assignment
By now, you have learned a bit about regression models. In this assignment, you will practice:

* Performing linear regression
* Performing multiple regression

**This assignment is due roughly two weeks from now at 11:59 pm on Friday, March 25.** It should be uploaded into the "Homework Assignments" submission folder for Homework #4.  Submission instructions can be found at the end of the notebook. **There are 104 standard points possible in this assignment with 5 additional bonus points possible. The distribution of points can be found in the section headers**.


---
## Part 1: Git (12 points)

For this assignment, you're going to add it to the `cmse202-s22-turnin` repository you created in class so that you can track your progress on the assignment and preserve the final version that you turn in. In order to do this you need to

**&#9989; Question 1.1 (1 point)**: Navigate to your `cmse202-s22-turnin` **local** repository and create a new directory called `hw-04` and copy this notebook in that new directory.

``` bash
mkdir hw-04

```


&#9989; **Question 1.2 (1 point)**: Before you move on, put the command that your instructor should run to pull changes from  your repository in the markdown cell below.

``` bash
if already exists:
git pull
else:
git clone https://github.com/Rajmeet/cmse202-s22-turnin
```

&#9989; **Question 1.3 (2 points)**: Before you move on, create a new branch called `hw04_branch` and move into it. In the cell below put the command(s) to create a new branch and to checkout the new branch. (_Note_: your TA will be able to see if you have created the branch and its history).

``` bash
git checkout -b hw04_branch

```

&#9989; **Question 1.4 (3 points)** Check the status of your local `git`.

``` bash 
git status

```
Copy and paste below the output of the command.

``` bash
On branch main
Your branch is up to date with 'origin/main'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        hw-04/

nothing added to commit but untracked files present (use "git add" to track)
```

What is the name of the branch you are in ? 

``` bash
hw04_branch
```

**Important:** If you are still in the `main` branch you must switch to the new branch. 

&#9989; **Question 1.5 (3 points):**
Add your name and GitHub username to the top of the notebook, then add and commit **ONLY** the notebook.

``` bash
git commit -m "HW04 Added"
```

What is the commit message you used ?

``` bash
HW04 Added
```

&#9989; **Question 1.6 (1 point):** Before moving on. Check that the notebook you are working on is the correct one. Run the following cell. **Are you in the new folder you just created?** If not close this notebook and open the one in the `hw-04` folder


In [ ]:
!pwd

&#9989; **Question 1.7 (1 point):** Finally push the updated notebook to GitHub.

``` bash
git push --set-upstream origin hw04_branch
```


**Important**: Double check you've added your Professor and your TA as collaborators to your "turnin" repository (you should have done this in the previous homework assignment).

**Also important**: Make sure that the version of this notebook that you are working on is the same one that you just added to your repository! If you are working on a different copy of the notebook, **none of your changes will be tracked**!

If everything went as intended, the file should now show up on your GitHub account in the "`cmse202-s22-turnin`" repository inside the `hw-04` directory that you just created within the new branch `hw04-branch`.

Periodically, **you'll be asked to commit your changes to the repository and push them to the remote GitHub location**. Of course, you can always commit your changes more often than that, if you wish.  It can be good to get into a habit of committing your changes any time you make a significant modification, or when you stop working on the project for a bit.


---
## Part 2: Linear Regression (56 points)

<img src="https://cdn.mos.cms.futurecdn.net/E3WFeLCEMJZvB9AFzv84TS.jpg" width=600px style="margin-left: 20px" alt="Image credit: https://blogs.acu.edu/">



### Part 2.1 Exploratory Data Analysis (33 points)
In this part of the homework, we will try to predict the price of diamonds.

&#9989;  **Question 2.1.1 (4 points)**: 

1. **Do This:** Download the diamonds dataset and write the command you used in the next cell

`https://raw.githubusercontent.com/msu-cmse-courses/cmse202-S22-data/main/data/diamonds.csv`
    
2. **Do This:** Load the data in this file into a `Pandas` dataframe and use the first column of the dataset as the index
3. **Do This:** Display the first *eight* rows of the dataframe. 

In [ ]:
!curl https://raw.githubusercontent.com/msu-cmse-courses/cmse202-S22-data/main/data/diamonds.csv -O

In [2]:
# Run this cell before moving on
%pylab
%matplotlib inline
import pandas as pd
import seaborn as sns

sns.set_context("talk")

Using matplotlib backend: agg
Populating the interactive namespace from numpy and matplotlib


In [ ]:
df = pd.read_csv("diamonds.csv")
df.head(8)

In [ ]:
df.dtypes, df.shape

&#9989;  **Question 2.1.2 (3 points)**: Answer the following questions in the next markdown cell

1. How many columns are there?

2. What `type` are the entries in this dataset ? For example, are they `int`, `float` ? You may need to use one of `pandas.DataFrame`'s methods to find the answer to this.

3. How many rows are in the dataset ? 

<font size=+3>&#9998;</font> 
1. There are 11 columns
2. Most of them are float except the price that's int
3. There are 53940 rows.

&#9989;  **Question 2.1.3 (1 points)**: Print out a quick statistical summary

In [ ]:
df.describe()

In [ ]:
df.loc[df.price.idxmax()]

&#9989;  **Question 2.1.4 (4 points)**: Answer the following questions in the next markdown cell.

1. What is the highest price for a diamond ? 

2. What `cut`, `color`, `clarity` and `carat` correspond to the most expensive diamond? _Hint:_ You might have to use some code for this.

3. Does the carat of the most expensive diamond correspond to the highest value in the carat column? Does this surprise you ?

4. Why are there missing columns in the statistical summary?

<font size=+3>&#9998;</font>
1. 18823.
2. For the most expensive diamond the cut is Premium, the clarity is VS2 and the carat is 2.29.
3. Yes as I thought the the higher the carat value the higher would be the price.
4. There are missing columns because they of are type object and not numeric.

&#9989;  **Question 2.1.5 (3 points)**: As you can see there are several values for `cut`, `clarity`, `color`. You could google what does values are. However, we want to test your coding skills, so in the next cell write some code and *print* the unique values for the `cut`, `clarity`, `color` columns

In [ ]:
print(df['cut'].unique())
print(df['clarity'].unique())
print(df['color'].unique())

Our goal is to predict the price of a diamond given its characteristics. However, up to now we have only investigated the dataset and still haven't found relationships between the columns. Let's do that now.

&#9989;  **Question 2.1.6 (3 points)**: First let's drop the `x`, `y`, `z` columns. They are not too important as they are used to calculate the column `depth` and `table`. Therefore, the latter two are more useful. 

Then, using `seaborn` make a pairplot of the dataset without the `x`, `y`, `z` columns. Use the `hue` option to differentiate between different types of diamonds.

In [ ]:
df = df.drop(['x','y','z'], axis=1)

In [ ]:
sns.pairplot(df,hue='color')

&#9989;  **Question 2.1.7 (9 points)**: Answer the following questions **without the use of any code!!!**.

1. What columns did you use in `hue` ? 

2. Why did you choose this value ?

3. Is there any pattern or specific information that you can extract from this plot? 

4. For each of the following pairs estimate a correlation value. 

    Choices are [greater than 0.5, less than -0.5, approximately 0] 
    
        a. table vs price  
        b. depth vs price
        c. carat vs price   
        d. depth vs table
        
5. If you were to use linear regression what columns would you choose for your X, and Y ?

<font size=+3>&#9998;</font>
1. I used the color column for hue.
2. I choose as every dimaond can have varing dimensions and there would be a lot to consider when classifying diamonds. Colors are easily distinguished.
3. It's really hard to find a specific information from this plot because of the data being very scattered.
4. 
5. 

&#9989; **Question 2.1.8 (6 points)**: From experience we know that the price of a diamond increases with its carats, but the dataset gives us more information and we would like to know what other factors influence the price. The pairplot above does not tell us much about the correlation between non-numeric columns so we need to find them ourselves.

The goal of this question is to find the characteristic with the highest correlation between price and carat. 
In other words, 

1. What value of `cut` has the highest correlation between `price` and `carat` ?

2. What value of `color` has the highest correlation between `price` and `carat` ?

3. What value of `clarity` has the highest correlation between `price` and `carat` ?

4. Which value of the above ones is the highest? 

**HINT:** This question requires a bit of ingenuity. Printing all the correlation matrices and looking the highest value is allowed, BUT you will get only half the points. You could definitely use this method to get insights, but can you write code that does not require me to read all those matrices? After all this class requires you to come up with ways to find the answer.

In [ ]:
df.groupby('cut')[['price','carat']].corr()

In [ ]:
df.groupby('color')[['price','carat']].corr()

In [ ]:
df.groupby('clarity')[['price','carat']].corr()

<font size=+3>&#9998;</font> 
1. Ideal cut has the highest correlation between price and carat.
2. J color has the highest correlation between price and carat
3. SI1 clarity has the highest correlation between price and carat.

### Part 2.2 Linear Regression with `statsmodels` (23 points).

The goal of part 2.1 was to find the independent and dependent variables for our linear regression. Now that we have them let's explore some linear regression.


&#9989;  **Question 2.2.1 (3 points)**: Based on your answer from the above question 

1. use a mask to select the data with the highest correlation value overall
2. make a scatter plot with `price` on the Y axis and `carat` on the X axis
3. Don't forget to label the axis!!!


In [ ]:
SI1_mask = df['clarity'] == 'SI1'
SI1_cut = df[SI1_mask]
np.corrcoef(SI1_cut['price'], SI1_cut['carat'])

In [ ]:
plt.scatter(SI1_cut['carat'], SI1_cut['price'])
plt.xlabel("Carat Size")
plt.ylabel("Price")

In [ ]:
plt.scatter(SI1_cut['price'], SI1_cut['carat'])
plt.yscale('log')
plt.xlabel('carat')
plt.ylabel('price')

In [ ]:
plt.scatter(SI1_cut['price']**0.5, SI1_cut['carat'])
plt.xlabel('carat')
plt.ylabel('price')

&#9989;  **Question 2.2.2 (3 points)**: Answer the following question

1. Does the plot look linear?

2. Does the plot look linear if you plot the `log` of `price` ? 

3. Does the plot look linear if you plot the `sqrt` of `price` ?

<font size=+3>&#9998;</font> 
1. Yes this plot looks linear.
2. No it does not look linear if we take the log price.
3. Yes it looks linear.

&#9989;  **Question 2.2.2 (6 points)**: Based on your previous answer. 

1. Use `statsmodels` to perform linear regression on the most linear plot.

2. Print a summary of your results

***Hint:*** Look over your PCAs and ICAs 

In [ ]:
import statsmodels.api as sm
X = SI1_cut['carat']
y = SI1_cut['price']**0.5
X = sm.add_constant(X)
model = sm.OLS(y,X)
result = model.fit()
print(result.params)
result.summary()

&#9989;  **Question 2.2.3 (3 points)**: Based on your previous answer. 

1. What are the values of the slope and intercept?

2. What is the value of your Adj. R-squared ?

3. Looking at the summary, do you trust that a linear fit is a good choice ?

<font size=+3>&#9998;</font> 
1. The slope is about 60 and the intercept is 6.11
2. The value of Adj. R-squared is 0.960.
3. Yes as looking at the data we see that the plot is linear and using a linear fit would be a great choice.

&#9989;  **Question 2.2.4 (3 points)**: Use [`sm.graphics.plot_regress_exog`](https://www.statsmodels.org/dev/generated/statsmodels.graphics.regressionplots.plot_regress_exog.html) to plot your regression results. Make your figure 15 x 8.


In [ ]:
fig = plt.figure(figsize=(15,8))
sm.graphics.plot_regress_exog(result,'carat',fig=fig)
plt.tight_layout()
plt.show()

&#9989;  **Question 2.2.5 (3 points)**: Do you think that a linear regression is a valid model ? Justify your answer.

**HINT:** Look at what we have done in the ICA.



<font size=+3>&#9998;</font> Yes it is a valid model. We can say that as by looking the fitted linear line works well for the data.

---
### &#128721; STOP
**Pause to commit your changes to your Git repository! (2 points)**

Take a moment to save your notebook, commit the changes to your local `git` repository using the commit message "Part 2 complete", and push the changes to GitHub.

---

## Part 3: Mulitple Regression (33 points + 5 BONUS points)

In this part of the homework, you will work with data from the California Housing Prices dataset. Our goal is to use ordinary least squares to design a linear model to fit the median house value in a California census district given eight features. We will try both a linear model using all the provided features, as well as a reduced linear model which only uses a subset of the features. 

&#9989;  **Question 3.1 (1 points)**: **Do This:** Download the file `housing.csv` from the link below, and save it into the same directory as your notebook. **Then, in the cell below, put the command line command(s) you used to download the file.** If you did not use a command line tool to download the file, write down the command(s) that would have downloaded the file.

`https://raw.githubusercontent.com/msu-cmse-courses/cmse202-S22-data/main/data/housing.csv`

``` bash
curl https://raw.githubusercontent.com/msu-cmse-courses/cmse202-S22-data/main/data/housing.csv -O
```

&#9989;  **Question 3.2 (2 points)**: **Do This:** To get started, read in the `housing.csv` dataset into a Pandas DataFrame, and then display some of its rows.

In [15]:
import statsmodels.api as sm
housing = pd.read_csv('housing.csv')
housing.head(10)

longitude  latitude  housing_median_age  avg_rooms  avg_bedrooms  \
0    -122.23     37.88                  41   6.984127      1.023810   
1    -122.22     37.86                  21   6.238137      0.971880   
2    -122.24     37.85                  52   8.288136      1.073446   
3    -122.25     37.85                  52   5.817352      1.073059   
4    -122.25     37.85                  52   6.281853      1.081081   
5    -122.25     37.85                  52   4.761658      1.103627   
6    -122.25     37.84                  52   4.931907      0.951362   
7    -122.25     37.84                  52   4.797527      1.061824   
8    -122.26     37.84                  42   4.294118      1.117647   
9    -122.25     37.84                  52   4.970588      0.990196   

   population  households  median_income  median_house_value ocean_proximity  
0         322         126         8.3252              452600        NEAR BAY  
1        2401        1138         8.3014              358500        NEAR BAY  
2         496         177         7.2574              352100        NEAR BAY  
3         558         219         5.6431              341300        NEAR BAY  
4         565         259         3.8462              342200        NEAR BAY  
5         413         193         4.0368              269700        NEAR BAY  
6        1094         514         3.6591              299200        NEAR BAY  
7        1157         647         3.1200              241400        NEAR BAY  
8        1206         595         2.0804              226700        NEAR BAY  
9        1551         714         3.6912              261100        NEAR BAY

&#9989;  **Question 3.3 (2 points)**: You should notice that the DataFrame has a non-numerical feature called "ocean_proximity". There are also a few rows with `NaN` values, although you may not see them in the few rows that were displayed. **Do This:** **Drop the "ocean_proximity" column from the dataframe, and drop all the rows with `NaN` values.** Note that we will not use the "ocean_proximity" column or any of the rows with `NaN` values in this assignment.

In [16]:
housing = housing.drop(['ocean_proximity'], axis=1)
housing = housing.dropna()
housing

longitude  latitude  housing_median_age  avg_rooms  avg_bedrooms  \
0        -122.23     37.88                  41   6.984127      1.023810   
1        -122.22     37.86                  21   6.238137      0.971880   
2        -122.24     37.85                  52   8.288136      1.073446   
3        -122.25     37.85                  52   5.817352      1.073059   
4        -122.25     37.85                  52   6.281853      1.081081   
...          ...       ...                 ...        ...           ...   
20635    -121.09     39.48                  25   5.045455      1.133333   
20636    -121.21     39.49                  18   6.114035      1.315789   
20637    -121.22     39.43                  17   5.205543      1.120092   
20638    -121.32     39.43                  18   5.329513      1.171920   
20639    -121.24     39.37                  16   5.254717      1.162264   

       population  households  median_income  median_house_value  
0             322         126         8.3252              452600  
1            2401        1138         8.3014              358500  
2             496         177         7.2574              352100  
3             558         219         5.6431              341300  
4             565         259         3.8462              342200  
...           ...         ...            ...                 ...  
20635         845         330         1.5603               78100  
20636         356         114         2.5568               77100  
20637        1007         433         1.7000               92300  
20638         741         349         1.8672               84700  
20639        1387         530         2.3886               89400  

[20433 rows x 9 columns]

&#9989;  **Question 3.4 (6 points)**: **Do This:** Using the `OLS()` method in `statsmodels.api`, make an ordinary least squares model that predicts "median_house_value" using the other features and display the `.summary()` of that process. **Be sure to use the `add_constant()` method to add a column of ones to the DataFrame before using the `OLS()` method so that your linear model includes a constant term.**

In [6]:
housing = sm.add_constant(housing)
x = housing['median_house_value'] 
y = housing.drop(columns=['median_house_value'])
# X = sm.add_constant(x)
model = sm.OLS(x,y)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:     median_house_value   R-squared:                       0.639
Model:                            OLS   Adj. R-squared:                  0.639
Method:                 Least Squares   F-statistic:                     4514.
Date:                Fri, 25 Mar 2022   Prob (F-statistic):               0.00
Time:                        18:28:09   Log-Likelihood:            -2.5677e+05
No. Observations:               20433   AIC:                         5.136e+05
Df Residuals:                   20424   BIC:                         5.136e+05
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const              -3.558e+06   6.35e+04    -56.021      0.000   -3.68e+06   -3.43e+06
longitude          -4.223e+04    725.615    -58.205      0.000   -4.37e+04   -4.08e+04
latitude           -4.213e+04    692.323    -60.854      0.000   -4.35e+04   -4.08e+04
housing_median_age  1139.3771     43.181     26.386      0.000    1054.739    1224.015
avg_rooms          -6625.2745    573.271    -11.557      0.000   -7748.932   -5501.617
avg_bedrooms        4.571e+04   2737.024     16.699      0.000    4.03e+04    5.11e+04
population           -41.2977      1.036    -39.865      0.000     -43.328     -39.267
households           134.5370      3.087     43.583      0.000     128.486     140.588
median_income       4.156e+04    406.403    102.256      0.000    4.08e+04    4.24e+04
==============================================================================
Omnibus:                     4964.002   Durbin-Watson:                   0.988
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            20949.365
Skew:                           1.141   Prob(JB):                         0.00
Kurtosis:                       7.404   Cond. No.                     2.52e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.52e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

&#9989;  **Question 3.5 (6 points)**: **Answer These Questions:** 
1. Based on the output of the OLS summary, does this linear model appear to be a good fit? 
2. Which of these features appear to be "significant" in predicting the "median_house_value"?

<font size=+3>&#9998;</font> 
1. Yes it seems to be a good fit. I can say that because of the R-squared value is high and the p value is low for the data. 
2. Features can be said that theyt are significant by looking at their P values. A Pvalue less than 0.05 is statistically significant. In our case all the features have a P value of 0 and hence all are significant. 

&#9989;  **Question 3.6 (4 points)**: A student from U of M managed to enroll in this class (don't ask us how) and is working on this homework. They got the following output when they used the `OLS()` method in `statsmodels.api`. 

<img src="https://i.ibb.co/pnvcKQ2/Perfect-Model.png" width="400px" border="0">

When you compare outputs, the Wolverine tells you "My linear model is better than yours because my linear model has a larger $R^2$-value than your model. I guess Wolverines are way better than Spartans at data science."

**Answer These Questions:** 
1. How did they get a linear model with a near perfect $R^2$-value? 
2. Is their model actually better than yours? 

Justify your answers.

<font size=+3>&#9998;</font> 
1. They get a perfect $R^2$ value because they have the median_house_value also added as a feature.
2. No the model is not better than mine. We can see that because of their high P values, which tells that the features are not significant. 

&#9989;  **Question 3.7 (6 points)**: In the output of the OLS summary, you should have seen a note that says something like 

"The condition number is large, [[number]]. This might indicate that there are strong multicollinearity or other numerical problems." 

Multicollinearity is a statistical phenomenon where some of the features in a model can be linearly predicted using some of the other features in the model. In other words, the features in the model are somewhat redundant. Hence, even if each feature may be deemed significant, it may still be possible to form a "reduced" model using a smaller number of features. 

**Do This:** See if you can design a second linear model which uses a subset of the eight features to predict the "median_house_value", and fits the data comparably well as the first linear model you designed in Question 3.4. You can choose this subset either by trial and error, or by any other method you'd like.

In [40]:
housing = sm.add_constant(housing)
x = housing['median_house_value']
y = housing.drop(columns=['median_house_value', 'avg_rooms', 'avg_bedrooms'])
model = sm.OLS(x,y)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:     median_house_value   R-squared:                       0.632
Model:                            OLS   Adj. R-squared:                  0.632
Method:                 Least Squares   F-statistic:                     5847.
Date:                Fri, 25 Mar 2022   Prob (F-statistic):               0.00
Time:                        19:36:34   Log-Likelihood:            -2.5696e+05
No. Observations:               20433   AIC:                         5.139e+05
Df Residuals:                   20426   BIC:                         5.140e+05
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const              -3.533e+06   6.15e+04    -57.468      0.000   -3.65e+06   -3.41e+06
longitude          -4.244e+04    697.445    -60.844      0.000   -4.38e+04   -4.11e+04
latitude           -4.271e+04    653.267    -65.386      0.000    -4.4e+04   -4.14e+04
housing_median_age  1137.0270     43.442     26.173      0.000    1051.877    1222.177
population           -43.5718      1.036    -42.061      0.000     -45.602     -41.541
households           141.0042      3.075     45.857      0.000     134.977     147.031
median_income       3.796e+04    271.798    139.654      0.000    3.74e+04    3.85e+04
==============================================================================
Omnibus:                     5303.846   Durbin-Watson:                   0.935
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            20764.348
Skew:                           1.248   Prob(JB):                         0.00
Kurtosis:                       7.262   Cond. No.                     2.41e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.41e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

&#9989;  **Question 3.8 (4 points)**: **Answer This:** How did your reduced linear model fit the data compared to the full linear model you created in Question 3.4? Give some quantitative justification for this answer.

<font size=+3>&#9998;</font> 
We are trying to maximize the R^2 value but find a good fit for the model. After trial and error and removing features like avg_rooms and avg_bedrooms, we are able to get a value of 0.632 which is close to the value we had with the when they were in (difference of 0.07). Reducing the data can be helpful as it will allow us to ignore some of the features that don't have that high of an impact on the predicted model.  

&#9989;  **BONUS Question 3.9 (5 points)**: In Question 3.4, the directions specified to fit a linear model **with a constant term**. For fun, we fit a linear model without a constant term as well. Here is a comparison of the OLS summaries for the full linear model with a constant term (left) and without a constant term (right).

<img src="https://i.ibb.co/mHJmtWQ/Comparison.png" width="800px" border="0">

You'll notice that the $p$-value for the constant term is very small. So one might argue that the constant term is a "significant" component of the model, and thus the model with the constant term is better. However, the $R^2$-value is much higher for the linear model without a constant term, and so one could also argue that the model without a constant term is better. 

**Answer This:** Can you explain what seems like a contradiction here? Which model (with or without a constant) do you think is more appropriate for predicting the "median_house_value", and why?

<font size=+3>&#9998;</font> I think a good answer for this would be asking the question why are constants import in the first place. If we consider the equation $y=mx+b$. If we set x = 0 we can predict that the value is going to be B. But if we make the constant term b = 0, then we get $y = mx$. This would be mean that y =0 whhich means that we are assuming that the line is going to pass through the center which is not always true. Moreover because the model now just depends on x values primarly, it will predict values that are very high or very low. The constant hence forms a baseline for doing such prediction and removes any unecessary bias when predicting the data.

---
### &#128721; STOP
**Pause to commit your changes to your Git repository! (2 points)**

Take a moment to save your notebook, commit the changes to your local `git` repository using the commit message "Part 3 complete", and push the changes to GitHub.

---

## Part 1. Continued (3 points)


It is now time to merge your changes to the main branch and push

&#9989; **Question 1.8 (1 point):** Switch to your main branch

```bash
# Put the command(s) to switch to your main branch
```

&#9989; **Question 1.9 (1 point):** Merge your hw04-branch

```bash
# Put the command(s) to merge your branch
```

&#9989; **Question 1.10 (1 point):** Finally push the changes to GitHub

```bash
# Put the command(s) to push your new main branch
```

---
## Assignment wrap-up

Please fill out the form that appears when you run the code below.  **You must completely fill this out in order to receive credit for the assignment!**

In [ ]:
from IPython.display import HTML
HTML(
"""
<iframe 
	src="https://forms.office.com/Pages/ResponsePage.aspx?id=MHEXIi9k2UGSEXQjetVofVWe89dpqk1HpKW7eyyv-IhUOU5NSUtUUFNFMTBWVEdKQUkzR1ZPWllHWi4u" 
	width="800px" 
	height="600px" 
	frameborder="0" 
	marginheight="0" 
	marginwidth="0">
	Loading...
</iframe>
"""
)

## Congratulations, you're done!

Submit this assignment by uploading it to the course Desire2Learn web page.  Go to the "Homework Assignments" folder, find the dropbox link for Homework #4, and upload it there.

![Moira2](https://media.giphy.com/media/d1E2HnwywoTkES08/giphy.gif)

&#169; Copyright 2022,  Department of Computational Mathematics, Science and Engineering at Michigan State University